src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -226,12 +234,51 @@ private Map<String, String> mergeLabels(
         return merged;
     }
 
+    private List<AbstractNode> modifyChildrenWithJvmIds(
+            Collection<? extends AbstractNode> children) {
+        List<AbstractNode> modifiedChildren = new ArrayList<>();
+        for (AbstractNode child : children) {
+            if (child instanceof TargetNode) {
+                ServiceRef ref = ((TargetNode) child).getTarget();
+                try {
+                    ref = jvmIdHelper.get().resolveId(ref);
+                } catch (Exception e) {
+                    logger.warn("Failed to resolve jvmId for node {}", child.getName());
+                    logger.info(e.getCause().toString());
```

Comment:
Likewise, I'm also hesitant about allowing these to pass and storing them into the database with null JVM IDs. Since we're already reaching out to the target to get the ID we may as well take the opportunity to be defensive about the data model and reject updates that attempt to publish targets we don't think are valid.

Reply:
Mmm... what if the error is of SecurityException i.e. invalid credentials or ssl error? 

> Scrolling through the rest of the diff, I do see that the `JvmIdHelper` has been added to the `PlatformClient` implementations that hook into the `BuiltInDiscovery`. I think this can be refactored as I was explaining above and all that shared logic around resolving the hash ID for a ServiceRef can be pulled into `DiscoveryStorage`, rather than being handled in each platform client plus the custom targets API handler - they're all pushing that data into the same storage anyway.
> 
> If the hash ID resolution fails then the storage can examine the failure reason:
> 
>     1. If it's a JMX SSL failure, just log a warning and complete the plugin update with a null ID for that ServiceRef
> 
>     2. If it's a JMX Auth failure, same
> 
>     3. If it's a network exception like the hostname specified in the connectUrl could not be resolved, or the connection timed out, etc. then fail the update. Either by removing that single element that could not be reached, or probably just by throwing an exception and letting that bubble up to whatever plugin tried to publish this update.

You said it's fine before, so I am just confirming because I think what I've already written accounts for that. Unless I'm missing something.

Reply:
You're right, what you've written does implement the behaviour I described before, but I'm second guessing what I said...

Give me a bit of time to think this through and figure out which way I want it to go. On the one hand, the SSL/Auth failures tell us that something really is there but we just aren't able to complete the task, so it seems valid to allow the update and include this in the discovery database. On the other hand, the ID is meant to be a unique identifier for an instance of a target, so nullable IDs seem likely to break things in unexpected ways in the future.

I'm also wondering to do about IDs for non-JMX Cryostat Agents. I've been thinking that such a Cryostat Agent can register itself with a `connectUrl` starting with `cryostat-agent://` rather than the typical `service:rmi:jmx://`. Cryostat would not be able to connect to the agent over JMX, but in this case we can assume and assert that the same agent must also be registered as a Discovery plugin (publishing only itself during the update we're processing now), which means we can enforce that the `connectUrl` has the `cryostat-agent` URI scheme rather than `http`, but uses the same host/IP:port as it used for its `callbackUrl`. But, since we want everything to have IDs and we can't necessarily connect back to these agents over JMX to retrieve the ID, we would need to allow the agent to also supply its own ID.

Or maybe we implement an alternate `JFRConnection` that is actually a facade overtop of HTTP connections to an Agent, and we implement a subset of capabilities on the Agent's integrated webserver. One of the supported operations could be retrieving the ID.

Reply:
> SSL/Auth
Hmm... not sure but maybe just preventing them from being added to the db may be better since I assume if we can't connect to the target for the ID because of the those errors, then we aren't able to perform any actions anyways on them? I'm wondering about the potential user cases for if someone doesn't provide SSL or something, but still wants to persist that target to the db...

> non-JMX Cryostat Agents
I think I like having the alternate JFRConnection over the other method because we wouldn't have to force extra functional compatibility from outside agents, then the fallback is always just some http attributes about the agent and maybe some other id given by the agent itself -> and making some id out of that, perhaps. So sort of both ideas at once. The cryostat-agent URI scheme maybe can then be used to identify using between the two different implementations to choose which "JFRConnection" to use.

Reply:
> Hmm... not sure but maybe just preventing them from being added to the db may be better since I assume if we can't connect to the target for the ID because of the those errors, then we aren't able to perform any actions anyways on them? I'm wondering about the potential user cases for if someone doesn't provide SSL or something, but still wants to persist that target to the db...

It might be useful for the user to still be able to see that they exist but are not connectable - we do that right now in Cryostat anyway, you can see the SSL-enabled or JMX-Auth-enabled targets in the API responses or web UI, and you just get a prompt when you try to perform an action on them if the SSL cert is untrusted or there are no good auth credentials stored. If we drop those targets from plugin publishes or reject the update entirely then the targets would never appear in the API responses or UI and the user would lose that ability to interactively provide credentials via X-JMX-Authorization passthrough headers etc. It'd be a pretty fundamental change to the interactive use cases for Cryostat. It seems like the "right", pure way to handle things, but it might just be too overly defensive to feel understandable to end users.

I think I'm leaning toward keeping the behaviour that I outlined before and which you've implemented here. Be specifically lenient about the SSL/Auth failure scenarios and allow null IDs only in that case, which indicates a "known" target that we are unable to connect to. if the problem was the SSL cert then it's taken care of when the user adds the cert and restarts Cryostat. If the problem was JMX Auth then the nulled ID should eventually get updated when something triggers Cryostat to retry connecting to that target. We wouldn't have that behaviour right now, though, because the null ID would be there from the time of the plugin update and would remain there until something caused the DiscoveryStorage to re-evaluate that target. We would need some hook to update those IDs in the DiscoveryStorage database after the fact, for example if the credentials are added late and an automated rule is triggered and causes Cryostat to successfully connect to the target.

Starting to look like we'll need that broken out relationship between Targets in the database as their own entities, and the discovery subtrees referencing those. This would allow us to update the JVM ID on a Target record individually, without having to figure out which discovery plugin subtree provided the target, update the one node, rebuild the tree, and store the subtree again.

>  think I like having the alternate JFRConnection over the other method because we wouldn't have to force extra functional compatibility from outside agents, then the fallback is always just some http attributes about the agent and maybe some other id given by the agent itself -> and making some id out of that, perhaps. So sort of both ideas at once. The cryostat-agent URI scheme maybe can then be used to identify using between the two different implementations to choose which "JFRConnection" to use.

Agreed, I think this imaginary "CryostatAgentConnection" that can do some subset of the usual JFRConnection things is probably the best way to go. I think we could pretty easily implement this within Cryostat by hooking it into the `TargetConnectionManager` and checking the URI scheme. If it's `cryostat-agent://` then we build a CryostatAgentConnection, otherwise we fall back to the current behaviour. The CryostatAgentConnection would just fire off HTTP requests to a small handful of API endpoints exposed by the Cryostat Agent to allow for things like retrieving a JVM ID (which could just be a UUID generated by the Cryostat Agent's JVM and held for the lifecycle), and maybe in the future also for listing, starting/stopping/deleting, and retrieving active flight recordings.

Reply:
> It might be useful for the user to still be able to see that they exist but are not connectable - we do that right now in Cryostat anyway, you can see the SSL-enabled or JMX-Auth-enabled targets in the API responses or web UI, and you just get a prompt when you try to perform an action on them if the SSL cert is untrusted or there are no good auth credentials stored. If we drop those targets from plugin publishes or reject the update entirely then the targets would never appear in the API responses or UI and the user would lose that ability to interactively provide credentials via X-JMX-Authorization passthrough headers etc. It'd be a pretty fundamental change to the interactive use cases for Cryostat. It seems like the "right", pure way to handle things, but it might just be too overly defensive to feel understandable to end users.
> 

Ah okay, that makes sense. 

>  without having to figure out which discovery plugin subtree provided the target, update the one node, rebuild the tree, and store the subtree again.
> 

Funny thing is that I've done this and added the hook for credentials being added to trigger a DiscoveryStorage.updateNode call in this pr already :-) . But it is definitely not the most efficient. Having a Target database like the the Credentials and Plugins may be nice...

> Agreed, I think this imaginary "CryostatAgentConnection" that can do some subset of the usual JFRConnection things is probably the best way to go. I think we could pretty easily implement this within Cryostat by hooking it into the `TargetConnectionManager` and checking the URI scheme. If it's `cryostat-agent://` then we build a CryostatAgentConnection, otherwise we fall back to the current behaviour. The CryostatAgentConnection would just fire off HTTP requests to a small handful of API endpoints exposed by the Cryostat Agent to allow for things like retrieving a JVM ID (which could just be a UUID generated by the Cryostat Agent's JVM and held for the lifecycle), and maybe in the future also for listing, starting/stopping/deleting, and retrieving active flight recordings.

Sounds good and makes sense to me, what should be the plan for now then?

Reply:
> Funny thing is that I've done this and added the hook for credentials being added to trigger a DiscoveryStorage.updateNode call in this pr already :-) . But it is definitely not the most efficient. Having a Target database like the the Credentials and Plugins may be nice...

Oh nice. We can probably move forward with what you already have then, so long as we get to the Target database stuff before the next release since it should be a rather large real-world performance improvement to do it that way.

> Sounds good and makes sense to me, what should be the plan for now then?

I have a bunch of other stuff on my plate for now so I'll come back around and re-review this PR. The next step after would be to pick up work on the Cryostat Agent again. Its implementation of a Discovery Plugin is pretty well done but could still probably use a bit of polishing.

The next thing it should have is the JFR Harvester that collects JFR data and periodically pushes it to Cryostat. That's already working to an extent, but we need a good way to identify this agent push as a sort of sourceTarget so that the data can go into the archives corresponding to the specific agent instance. This does tie in to what we're talking about with the CryostatAgentConnection since this could be what we use to pull an ID from the agent, and the agent can also use that as its sourceTarget path parameter when it's pushing JFR files, so maybe we implement the CryostatAgentConnection part first and then finish off work on the JFR Harvester part last.